### PR TITLE
Plain console only prints errors on stdout when console is attached

### PR DESCRIPTION
**Pull request by ghale**
This changes the behavior of the plain console such that it only redirects error messages to stdout when a console is attached to both stdout and stderr.  Otherwise, it sends error messages to stderr.  This allows error messages to be redirected to a file or a separate file handle when necessary, while the user-centric experience when a console is attached is that error messages are properly grouped on stdout with the rest of the output.

### CODE REVIEW

File: subprojects/logging/src/main/java/org/gradle/internal/logging/sink/PlainConsoleDispatchingListener.java [hunk 1]
```diff
@@ -42,9 +44,13 @@ public void onOutput(OutputEvent event) {
             // TODO - should attempt to flush the output stream prior to writing to the error stream
             stderrChain.onOutput(event);
         } else {
-            // Write anything that isn't the build failure report to the output stream
+            // Write anything that isn't the build failure report to the output stream if there is a console attached.
             // This is to avoid interleaving of error and non-error output generated at approximately the same time, as the process stdout and stderr may be forwarded on different pipes and so ordering is lost
-            stdoutChain.onOutput(event);
+            if (redirectStderr) {
```
Comment (adammurdoch) [inline subprojects/logging/src/main/java/org/gradle/internal/logging/sink/PlainConsoleDispatchingListener.java]: should also remove the stuff above that checks the category and for to stderr, on line 40, as the new logic replaces this old (and hacky) logic.

File: subprojects/logging/src/integTest/groovy/org/gradle/internal/logging/console/taskgrouping/AbstractBasicGroupedTaskLoggingFunctionalTest.groovy [hunk 2]
```diff
@@ -93,7 +93,12 @@ abstract class AbstractBasicGroupedTaskLoggingFunctionalTest extends AbstractCon
         succeeds('log')
 
         then:
-        result.groupedOutput.task(':log').output == "Standard out 1\nStandard err 1\nStandard out 2\nStandard err 2"
+        if (consoleAttached && stderrAttached) {
```
Comment (adammurdoch) [inline subprojects/logging/src/integTest/groovy/org/gradle/internal/logging/console/taskgrouping/AbstractBasicGroupedTaskLoggingFunctionalTest.groovy]: I think we're also missing some coverage that shows that the error output for two tasks running in parallel is not interleaved, but is instead grouped by task.

File: subprojects/logging/src/integTest/groovy/org/gradle/internal/logging/console/taskgrouping/plain/PlainConsoleAttachedToStdoutBuildSrcGroupedTaskFunctionalTest.groovy [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.logging.console.taskgrouping.plain
+
+import org.gradle.api.logging.configuration.ConsoleOutput
+import org.gradle.internal.logging.console.taskgrouping.AbstractConsoleBuildSrcGroupedTaskFunctionalTest
+
+class PlainConsoleAttachedToStdoutBuildSrcGroupedTaskFunctionalTest extends AbstractConsoleBuildSrcGroupedTaskFunctionalTest {
```
Comment (adammurdoch) [inline subprojects/logging/src/integTest/groovy/org/gradle/internal/logging/console/taskgrouping/plain/PlainConsoleAttachedToStdoutBuildSrcGroupedTaskFunctionalTest.groovy]: I wonder if it's time for a test runner for these console tests that runs each test using each combination of (console type x console attachment), instead lots of empty subclasses? Not sure.

File: subprojects/logging/src/main/java/org/gradle/internal/logging/sink/PlainConsoleDispatchingListener.java [hunk 4]
```diff
@@ -42,9 +44,13 @@ public void onOutput(OutputEvent event) {
             // TODO - should attempt to flush the output stream prior to writing to the error stream
```
Comment (adammurdoch) [inline subprojects/logging/src/main/java/org/gradle/internal/logging/sink/PlainConsoleDispatchingListener.java]: We should think about doing this or perhaps just add an issue to the backlog to make this better

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/ExecutionResult.java [hunk 5]
```diff
@@ -68,6 +68,13 @@
      */
     boolean hasErrorOutput(String expectedOutput);
 
+    /**
+     * Returns true when this result includes the given output log message. Considers only text in the main content of the build output and not in or following the build result message (use {@link #assertHasPostBuildOutput(String)} instead).
+     *
+     * @param expectedOutput The expected log message, with line endings normalized to a newline character.
+     */
+    boolean hasMainOutput(String expectedOutput);
```
[the remark this answers is not in the document]
Comment (ghale) [reply subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/ExecutionResult.java]: Yeah, this was _only_ here to support the delegation in `ErrorsOnStdoutOutputScrapingResult`.  I've just removed this and handled it differently.

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/InProcessGradleExecuter.java [hunk 6]
```diff
@@ -472,6 +473,11 @@ public boolean hasErrorOutput(String expectedOutput) {
             return outputResult.hasErrorOutput(expectedOutput);
         }
 
+        @Override
+        public boolean hasMainOutput(String expectedOutput) {
+            return outputResult.hasErrorOutput(expectedOutput);
```
[the remark this answers is not in the document]
Comment (ghale) [reply subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/InProcessGradleExecuter.java]: Ouch.  Removed along with the method.

File: subprojects/logging/src/main/java/org/gradle/internal/logging/sink/PlainConsoleDispatchingListener.java (hunk 1 above)
Comment (ghale) [reply subprojects/logging/src/main/java/org/gradle/internal/logging/sink/PlainConsoleDispatchingListener.java]: 👍  Done.

File: subprojects/logging/src/integTest/groovy/org/gradle/internal/logging/console/taskgrouping/plain/PlainConsoleAttachedToStdoutBuildSrcGroupedTaskFunctionalTest.groovy (hunk 3 above)
Comment (ghale) [reply subprojects/logging/src/integTest/groovy/org/gradle/internal/logging/console/taskgrouping/plain/PlainConsoleAttachedToStdoutBuildSrcGroupedTaskFunctionalTest.groovy]: Yeah, it is getting kind out of control.  I think we can probably defer to a follow up PR in an effort to get this one out the door.

File: subprojects/logging/src/integTest/groovy/org/gradle/internal/logging/console/taskgrouping/AbstractBasicGroupedTaskLoggingFunctionalTest.groovy (hunk 2 above)
Comment (ghale) [reply subprojects/logging/src/integTest/groovy/org/gradle/internal/logging/console/taskgrouping/AbstractBasicGroupedTaskLoggingFunctionalTest.groovy]: I've added this coverage to the multi-project test here: https://github.com/gradle/gradle/pull/5346/files#diff-f533c2c3d1ab352e8b1f9c3fef2f87c1.  The assertions are a little complex right now, but these should simplify once we get non-console output for rich/verbose consoles grouped.

**Pull request merged**